boards/arm/stm32/nucleo-f4x1re/src/stm32_ssd1306.c

```diff
@@ -0,0 +1,164 @@
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <debug.h>
+
+#include <nuttx/board.h>
+#include <nuttx/lcd/lcd.h>
+#include <nuttx/lcd/ssd1306.h>
+#include <nuttx/i2c/i2c_master.h>
+#include <nuttx/spi/spi.h>
+
+#include "stm32_i2c.h"
+#include "stm32_spi.h"
+
+#ifndef OK
+#  define OK 0
+#endif
+
+#define OLED_SPI_PORT 1
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct lcd_dev_s    *g_lcddev;
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: board_ssd1306_initialize
+ *
+ * Description:
+ *   Initialize and register the device. I2C version.
+ *
+ * Input Parameters:
+ *   busno - The I2C bus number
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_LCD_SSD1306_I2C
+int board_ssd1306_initialize(int busno)
+{
+  struct i2c_master_s *i2c;
+  const int devno = 0;
+
+  /* Initialize I2C */
+
+  i2c = stm32_i2cbus_initialize(busno);
+  if (!i2c)
+    {
+      lcderr("ERROR: Failed to initialize I2C port %d\n", busno);
+      return -ENODEV;
+    }
+
+  /* Bind the I2C port to the OLED */
+
+  g_lcddev = ssd1306_initialize(i2c, NULL, devno);
+  if (!g_lcddev)
+    {
+      lcderr("ERROR: Failed to bind I2C port %d to OLED %d\n", busno, devno);
+      return -ENODEV;
+    }
+  else
+    {
+      lcdinfo("Bound I2C port %d to OLED %d\n", busno, devno);
+
+      /* And turn the OLED on */
+
+      g_lcddev->setpower(g_lcddev, CONFIG_LCD_MAXPOWER);
+      return OK;
+    }
+}
+#endif
+
+/****************************************************************************
+ * Name: board_ssd1306_initialize
+ *
+ * Description:
+ *   Initialize and register the device. SPI version.
+ *
+ * Input Parameters:
+ *   busno - The SPI bus number
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_LCD_SSD1306_SPI
+int board_ssd1306_initialize(int busno)
+{
+  struct spi_dev_s *spi;
+  const int devno = 0;
+
+  /* Initialize SPI */
+
+  spi = stm32_spibus_initialize(busno);
+  if (!spi)
+    {
+      lcderr("ERROR: Failed to initialize SPI port %d\n", busno);
+      return -ENODEV;
+    }
+
+  /* Bind the SPI port to the OLED */
+
+  g_lcddev = ssd1306_initialize(spi, NULL, devno);
+  if (!g_lcddev)
```

Comment:
Optional
```suggestion
  if (g_lcddev == NULL)
```

Reply:
I'm going to write driver for e-ink SSD1600. I will take into consideration Yours suggestions.